keps/sig-api-machinery/4080-generic-controlplane/README.md

```diff
@@ -0,0 +1,768 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes factoring *kube-apiserver* to build on new staging repository 
+that consumes `k/apiserver` but has a bigger, carefully chosen subset of the 
+functionality of *kube-apiserver* such that it is reusable.
+
+The factoring will be progressive: we will start with a new repo that adds
+nothing to `k/apiserver`, and then progressively move generic functionality from
+*kube-apiserver* to the new repository. The new repo will be named 
+`k/generic-controlplane`.
+
+## Motivation
+
+A working kube-based control plane is more than just an apiserver component
+built on `k/apiserver`. It includes standard resources (depending on context
+namespaces, CRDs, RBAC, secrets, configmaps), and standard controllers (think 
+of garbage collection, namespace deletion, etc.). *kube-apiserver* today is a
+bundle of those resources with container orchestration, *kube-controller-manager* 
+equally for the corresponding controllers.
+
+Separating the generic parts from container orchestration will allow new 
+use-cases building upon `k/apimachinery` and `k/apiserver`, while keeping a 
+unified codebase and ecosystem, and by improving the factoring of 
+*kube-apiserver* for easier maintenance due to less complexity by clear layering.
+
+### Goals
+
+- As always: every PR transforms a working system into a working system, and 
+  PRs are of manageable size.
+- Improve factoring of *kube-apiserver* through layering on-top of 
+  `k/generic-controlplane`, reducing complexity through more explicit structure
+  and reduction of code in `k/k`.
+- `k/generic-controlplane` will provide
+  - a `sample-generic-controlplane` binary
+  - a modular, further customizable (in code) library suitable to build a 
+    working kube-based control plane without vendoring k/k.
+- `k/generic-controlplane` will (optionally) include the ability to define 
+  resources by CustomResourceDefinition objects.
+- `k/generic-controlplane` will be able to (optionally) delegate handling of 
+  some kinds of objects to another server, as directed by APIService objects.
+- `k/generic-controlplane` will allow customization (in code) of which generic
+  (native) resources like secrets, configmaps, admission webhooks, RBAC, etc. 
+  are served.
+- `k/generic-controlplane` will not include the definitions of the resources in 
+  Kubernetes for the management of containerized workloads.  For example, the 
+  excluded resources include: nodes, pods, daemonsets, ingresses, services, 
+  persistentvolumes.
+- `k/generic-controlplane` will allow, but not require all-in-one binaries, both 
+  in a *hyperkube*-like subcommand way, and in a all-in-one *k3s* like way.
```

Comment:
The decision to have separate binaries has aged extremely well.   We explicitly removed hyperkube and running controllers and apiservers in a single PID precludes separate tuning.  Combining as suggested in this goal looks like a mistake.

Reply:
Maybe important to point out (it's in the text): we do not provide a binary as part of this KEP as one-size-fits-all intentionally. So there is not decision about this shape of that necessary.